src/braket/default_simulator/simulator.py

```diff
@@ -386,6 +382,161 @@ def _observable_hash(observable: Observable) -> Union[str, dict[int, str]]:
+        qubit_map.
+
+        Args:
+            circuit (Circuit): The circuit containing the instructions.
+            qubit_map (dict): A dictionary mapping original qubits to new qubits.
+        """
+        for ins in circuit.instructions:
+            ins._targets = tuple([qubit_map[q] for q in ins.targets])
+
+    @classmethod
+    def _map_circuit_results(cls, circuit: Circuit, qubit_map: dict):
+        """
+        Maps the targets of each result in the circuit to the corresponding qubits in the qubit_map.
+
+        Args:
+            circuit (Circuit): The circuit containing the results.
+            qubit_map (dict): A dictionary mapping original qubits to new qubits.
+        """
+        for result in circuit.results:
+            if isinstance(result, (MultiTarget, OptionalMultiTarget)) and result.targets:
+                result.targets = [qubit_map[q] for q in result.targets]
+
+    @classmethod
+    def _map_jaqcd_instructions(cls, circuit: JaqcdProgram, qubit_map: dict):
+        """
+        Maps the attributes of each instruction in the JaqcdProgram to the corresponding qubits in
+        the qubit_map.
+
+        Args:
+            circuit (JaqcdProgram): The JaqcdProgram containing the instructions.
+            qubit_map (dict): A dictionary mapping original qubits to new qubits.
+        """
+        for ins in circuit.instructions:
+            cls._map_instruction_attributes(ins, qubit_map)
+
+        if cls.has_basis_rotation_instructions(circuit):
+            for ins in circuit.basis_rotation_instructions:
+                ins.target = qubit_map[ins.target]
+
+    @classmethod
+    def _map_instruction_attributes(cls, instruction, qubit_map: dict):
+        """
+        Maps the qubit attributes of an instruction to the corresponding qubits in the qubit_map.
+
+        Args:
+            instruction: The instruction whose qubit attributes need to be mapped.
+            qubit_map (dict): A dictionary mapping original qubits to new qubits.
+        """
+        if hasattr(instruction, "control"):
+            instruction.control = qubit_map[instruction.control]
+
+        if hasattr(instruction, "controls"):
+            instruction.controls = [qubit_map[q] for q in instruction.controls]
+
+        if hasattr(instruction, "target"):
+            instruction.target = qubit_map[instruction.target]
+
+        if hasattr(instruction, "targets"):
+            instruction.targets = [qubit_map[q] for q in instruction.targets]
+
+    @classmethod
+    def has_basis_rotation_instructions(cls, circuit):
+        """
+        Checks if the given circuit has basis rotation instructions.
+
+        Args:
+            circuit: The circuit to check for basis rotation instructions.
+
+        Returns:
+            bool: True if the circuit has basis rotation instructions, False otherwise.
+        """
+        return (
+            hasattr(circuit, "basis_rotation_instructions") and circuit.basis_rotation_instructions
+        )
```

Comment:
Is a separate method necessary? `Circuit` should always have a `basis_rotation_instructions` property; it just might be empty.